File: topics/mpi/practicals/01.MPI_Intro/fortran/2.hello_withrank_solution.f90
PROGRAM hello

!==============================================================!
!                                                              !
! This file has been written as a sample solution to an        !
! exercise in a course given at the CSCS Summer School.        !
! It is made freely available with the understanding that      !
! every copy of this file must include this header and that    !
! CSCS take no responsibility for the use of the enclosed      !
! teaching material.                                           !
!                                                              !
! Purpose: A program to try MPI_Comm_size and MPI_Comm_rank.   !
!                                                              !
! Contents: F-Source                                           !
!                                                              !
!==============================================================!

  USE MPI

  IMPLICIT NONE

  ! Don't forget to declare variables

  INTEGER :: ierror, rank, size

  CALL MPI_INIT(ierror)

  ! Get the rank of each process
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierror)

  ! Get the size of the communicator
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierror)

  ! Write code such that every process writes its rank and the size of the communicator,
  ! but only process 0 prints "hello world"
  IF (rank .EQ. 0) THEN
      WRITE(*,*) 'Hello world!'
  END IF

  WRITE(*,*) 'I am process', rank, ' out of', size

  CALL MPI_FINALIZE(ierror)

END PROGRAM

File: topics/mpi/solutions/01.MPI_Intro/cxx/2.hello_withrank_solution.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the CSCS Summer School.        *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * CSCS take no responsibility for the use of the enclosed      *
 * teaching material.                                           *
 *                                                              *
 * Purpose: A program to try MPI_Comm_size and MPI_Comm_rank.   *
 *                                                              *
 * Contents: C-Source                                           *
 ****************************************************************/

#include <stdio.h>
#include <mpi.h>


int main(int argc, char *argv[])
{
    /* declare any variables you need */
    int my_rank;
    int size;

    MPI_Init(&argc, &argv);

    /* Get the rank of each process */
    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    /* Get the size of the communicator */
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    /* Write code such that every process writes its rank and the size of the communicator,
     * but only process 0 prints "hello world*/
    if (my_rank == 0) {
        printf ("Hello world!\n");
    }

    printf("I am process %i out of %i.\n", my_rank, size);

    MPI_Finalize();
    return 0;
}
